src/python/tests/test_dynamic_api.py

```diff
@@ -0,0 +1,68 @@
+########################################################################
+#
+# Test the API for dynamically accessing fields of chemistry_data
+#
+#
+# Copyright (c) 2013, Enzo/Grackle Development Team.
+#
+# Distributed under the terms of the Enzo Public Licence.
+#
+# The full license is in the file LICENSE, distributed with this
+# software.
+########################################################################
+
+from pygrackle import chemistry_data
+
+from pygrackle.grackle_wrapper import _wrapped_c_chemistry_data
+
+def _test_dynamic_api(param_list, ok_vals, bad_vals, expected_output_type):
+    # test that the chemistry_data() class supports access to an attribute
+    # named for element in key_list.
+    #
+    # most other tests of chemistry_data() also implicitly test the dynamic_api
+
+    def _check_expected_type(param, val):
+        if not isinstance(val, expected_output_type):
+            raise AssertionError(
+                f"expected '{param}' to be an instance of "
+                f"'{expected_output_type.__name__}', not an instance of "
+                f"'{type(val).__name__}'"
+            )
+
+    obj = chemistry_data()
+    for param in param_list:
+        val = getattr(obj, param)
+        _check_expected_type(param, val)
+
+        for val_to_set in ok_vals:
+            setattr(obj, param, val_to_set)
+            val = getattr(obj, param)
+            _check_expected_type(param, val)
+
+            # check equality between val_to_set and val
+            if isinstance(val_to_set, str) and (expected_output_type == bytes):
+                assert val.decode('ascii') == val_to_set
+            else:
+                assert val == val_to_set
+
+        for val_to_set in bad_vals:
+            try:
+                setattr(obj, param, val_to_set)
+                raise ValueError(
+                    f"expected a TypeError when setting '{param}' to "
+                    f"{val_to_set!r}"
+                )
+            except TypeError:
+                continue
```

Comment:
Very minor suggestion, but I think this could be simplified as
```suggestion
            with assert_raises(TypeError):
                setattr(obj, param, val_to_set)
```
where `assert_raises` can be imported from `numpy.testing`.

Reply:
that's really neat - I definitely didn't know about that!